biogeophys/FatesHydroWTFMod.F90

```diff
@@ -0,0 +1,1136 @@
+          
+       else
+          write(fates_log(),*) 'TFS WRF was called for an ineligible porous media'
+          call endrun(msg=errMsg(sourcefile, __LINE__))
+          
+       end if !porous media
+       
+       ! Now lets smooth the result of capilary elastic with cavitation
+       
+       psi_cavitation = psi_sol
+       
+       b = -1._r8*(psi_capelast + psi_cavitation)
+       c = psi_capelast*psi_cavitation
+       
+       dcav_dth = dsol_dth
+       
+       dbdth = -1._r8*(dcapelast_dth + dcav_dth)
+       dcdth = psi_capelast*dcav_dth + dcapelast_dth*psi_cavitation
+       
+       dpsidth = 1._r8/(2._r8*quad_a2)*(-dbdth + 0.5_r8*((b*b - 4._r8*quad_a2*c)**(-0.5_r8)) * &
+            (2._r8*b*dbdth - 4._r8*quad_a2*dcdth))
+    end if
+
+    return
+  end function dpsidth_from_th_tfs
+
+  ! =====================================================================================
+
+  function ftc_from_psi_tfs(this,psi) result(ftc)
+
+    class(wkf_type_tfs) :: this
+    real(r8),intent(in) :: psi   !
+    real(r8)            :: ftc
+    real(r8)            :: psi_eff
+
+    psi_eff = min(0._r8,psi)
+
+    ftc = max(min_ftc,1._r8/(1._r8 + (psi_eff/this%p50)**this%avuln))
+
+  end function ftc_from_psi_tfs
+
+  ! ====================================================================================
+
+  function dftcdpsi_from_psi_tfs(this,psi) result(dftcdpsi)
+
+    class(wkf_type_tfs) :: this
+    real(r8),intent(in) :: psi
+    real(r8)            :: ftc      ! differentiated func
+    real(r8)            :: fx       ! portion of ftc function
+    real(r8)            :: dfx      ! differentiation of portion of func
+    real(r8)            :: dftcdpsi ! change in frac total cond wrt psi
+
+    ! Differentiate
+    ! ftc = 1._r8/(1._r8 + (psi/this%p50(ft))**this%avuln(ft))
+
+    if(psi>0._r8)then
+       dftcdpsi = 0._r8
+    else
+       ftc = 1._r8/(1._r8 + (psi/this%p50)**this%avuln)
+       if(ftc<min_ftc) then
+          dftcdpsi = 0._r8
+       else
+          fx  = 1._r8 + (psi/this%p50)**this%avuln
+          dfx = this%avuln*(psi/this%p50)**(this%avuln-1._r8) * (1._r8/this%p50)
+          dftcdpsi = -fx**(-2._r8)*dfx
+       end if
+    end if
+
+  end function dftcdpsi_from_psi_tfs
+
+  !-------------------------------------------------------------------------------!
+
+  subroutine solutepsi(th,rwc_fd,th_sat,th_res,pinot,psi)
+    !
+    ! !DESCRIPTION: computes solute water potential (negative) as a function of
+    !  water content for the plant PV curve.
+    !
+    ! !USES:
+    !
+    ! !ARGUMENTS
+
+    real(r8)      , intent(in)     :: th          ! vol wc       [m3 m-3]
+    real(r8)      , intent(in)     :: rwc_fd
+    real(r8)      , intent(in)     :: th_sat
+    real(r8)      , intent(in)     :: th_res
+    real(r8)      , intent(in)     :: pinot
+    real(r8)      , intent(out)    :: psi         ! water potential   [MPa]
+
+    psi = pinot*th_sat*(rwc_fd - th_res) / (th - th_sat*th_res)
```

Comment:
@xuchongang @JunyanDing @bchristo :

I worked it through too, and I think I understand now.

I rearranged a little bit, it looks like all the terms in the numerator and denominator are meant to be in units that match saturation fraction.

So if units of resid matches saturation fraction units: 

y = pinot * (rwcft - resid) / ( x/thetas - resid) 


If resid is in water content units (absolute, or same as thetas and theta):

y = pinot * (rwcft - resid/thetas) / ( x/thetas - resid/thetas) 

   = pinot * (rwcft*thetas - resid) / (x-resid)


I'm fine with either unit, just let me know what you all prefer and I will work it out.  We should also update the parameter file, and @JunyanDing noted that the derivative needs fixing too I think.

Reply:
I'm noting other places where we treat resid with units that are the same as water conent (not relativistic):

https://github.com/NGEET/fates/blob/c99c5bdf99cdb2ef6e8894b853b5b8ee9cb84627/biogeophys/FatesHydroWTFMod.F90#L823

Reply:
I can confirm what Chonggang noted -- indeed, anywhere in the old code there was the variable `resid`, it was as a fraction of `thetas`

Poor nomenclature, I can see how this misinterpretation happened. But the implication here is that anywhere in the current refactor that `th_res` represents a replacement of `resid` then that should be fixed.

Reply:
@rgknox and @bchristo , I would prefer using th_res as the same unit of thetas and theta. We only need to update the equations of derivative and water content, where old resid used.

Reply:
@rgknox and @xuchongang Here's my preference for how to deal with this:

- Change name of parameter in paramfile to rwc_r and note the relativistic RWC units there
- Calculate in `set_wrf_param_tfs` `th_res` as `thetas*rwc_r`
- Replace `th_res` in `solutepsi()`, `dpsolutepsi()`, `pressurepsi()`, and `dpressurepsi()` with `rwc_r`

Reply:
We could change the name of the parameter in the file in a later pull request, just not on this PR, as it will create incompatabilities for other users.  I think we have a PR coming soon to update a group of parameters.  That being said, the current parameter name seems suitably ambiguous, at least for right now, and we could change the descriptive text (in this PR) to make it clear that is the fraction of th_sat.   I will go with whatever you and @xuchongang decide.

Reply:
in an offline discussion, there was agreement that we will change the th_res parameter to have volumetric units, similar to th_sat